[Feature] 상품 기능 API 추가

Cass07 opened the pull request:
## 작업 내역 개요
Product 정보의 삽입/수정/삭제/조회 API, 여러 검색 조건을 선택적으로 사용 가능한 Product 목록 조회 API를 추가.  

- Product Entity를 생성하고, Member와의 연관관계 추가
	- 이에 따른 N+1 problem 을 방지하기 위해 list 조회 시 fetch join을 사용하도록 함
- Product CRUD 메소드 추가
- JPA Specification을 사용하는 동적 쿼리 추가
	- 보통은 쿼리가 복잡해지면 QueryDSL을 사용하게 되는데, 이후 복잡한 조회 쿼리가 더 추가되면 사용을 고려하는 것으로 하고 우선은 JPA만을 사용하여 구현하기로 결정함
- Product에 Facade 패턴을 사용하고, 이에 맞춰 사용할 수 있도록 MemberService에 Member 객체를 대신 리턴하는 조회 메소드를 일부 추가
- Product 이미지는 별도의 CRUD API를 구현할 예정
	- 이미지 업로드를 위해서는 form data로 받거나 URL 자체를 받아야 하는데, form data로 받으려면 다른 body 정보도 form data로 전송해야 하고, 이미지 처리에 시간이 걸릴 수 있기 때문에 API를 분리하는 것이 낫다고 판단함

Cass07 committed:
[Feature] 상품 등록/수정/삭제/조회 기능 추가

상품의 등록/수정/삭제/조회, 다양한 조회 쿼리를 사용 가능한 상품 목록 조회 추가
- Member 엔티티와 관계를 맺고 있는 Product 엔티티 추가
- 로그인한 사용자를 대상으로 상품 CRUD 메소드를 추가
- 여러 조회 쿼리를 선택적으로 사용할 수 있도록, 동적 쿼리를 사용하는 상품 목록 조회 메소드 추가 (JPA Specification)
- 상품 기능에 대해 Facade 레이어를 추가

Cass07 committed:
[Refactor] code smell 리팩토링

명명 규칙에 맞지 않는 변수명을 규칙에 맞도록 수정

Cass07 committed:
[Refactor] product 상품명의 동시성 문제

product의 상품명이 언제나 고유해질 수 있도록 발생할 수 있는 동시성 문제를 제어
- 기존 체크 로직은 유지
- name 컬럼을 Unique Key를 추가하여 DB에서 동시성 문제를 제어하도록 수정
- Product 객체의 save 및 update시, 바로 flush하도록 변경
- 데이터베이스의 무결성 제약조건을 위배할 때 발생하는 DataIntegrityViolationException 을 ExceptionHandler로 처리하도록 추가

f-lab-jd commented on src/main/java/dev/cass/knorda/api/product/facade/ProductFacade.java:
```diff
@@ -0,0 +1,63 @@
+	public ProductRegisterDto.RegisterResponse registerProduct(ProductRegisterDto.RegisterRequest registerRequest,
+		String memberName) {
+		Member member = memberService.findMemberByMemberName(memberName);
+		if (productService.isExistProductName(registerRequest.getProductName())) {
```
유효성 검사가 member 조회보다 먼저 수행되게 되면 불필요한 쿼리를 줄일 수 있습니다.

Cass07 replied:
유효성 검사 이후에 member를 조회하도록 순서를 변경하였습니다.

f-lab-jd commented on src/main/java/dev/cass/knorda/api/product/facade/ProductFacade.java:
```diff
@@ -0,0 +1,63 @@
+		return ProductFindDto.GetProductResponse.of(productService.findById(productId));
+	}
+
+	public ProductRegisterDto.RegisterResponse registerProduct(ProductRegisterDto.RegisterRequest registerRequest,
```
`@Transactional` 어노테이션을 붙이는 것은 아래 세개 중에 어떤 방법이 적당할까요? ㅎㅎ 
그리고 각각 어떤일이 벌어지나요~? 
1.  `Facade` 
2. `Service`
3.  둘 다

Cass07 replied:
기본적으로 모든 `JpaRepository`의 메소드는 Transactional이 적용됩니다
- `JpaRepository`를 구현했을 때, Spring Data JPA에서 알아서 `SimpleJpaRepository` 구현체를 생성해주기 때문
- 이 구현체애 `@Transactional(readOnly = true)가 붙어 있기 때문에 디폴트로 readOnly Transaction이 적용되어 있다  
또한, 기본적으로 다중 계층에서 Transactional 어노테이션이 여러 번 선언된 경우, 디폴트로는 상위 트랜잭션에 하위 트랜잭션이 들어가게 됩니다
- `@Transactional` 어노테이션에서 propagation(전파 속성)의 디폴트 값이 `DEFAULT`기 때문.
   - 내부의 여러 논리적인 트랜잭션을 묶어, 하나의 물리 트랜잭션으로 만듬
- 하위 트랜잭션을 별개의 물리적인 트랜잭션으로 분리하기 위해서는, `propagation = REQUIRES_NEW` 옵션을 지정해야 함  

즉 정리해보면,
1. facade에만 : facade 메소드 내에서 호출한 service 모든 메소드가 하나의 물리적인 트랜잭션으로 처리됨. 이 중 하나에서 문제가 발생하면 모두 롤백됨
2. service에만 : 각각의 service 메소드가 각자의 트랜잭션을 가짐. 이 중 하나에서 문제가 발생하면, 해당 커밋만 롤백되고, 이전 트랜잭션은 정상적으로 처리됨
3. 둘 다 : 1과 동일
![image](https://github.com/user-attachments/assets/79a76c2a-9f80-4a63-b36a-60228c80acdb)

원자성이 필요한 작업 단위를 어디서 결정할 것인지에 따라 달라지겠지만, 조회만 필요로 하는 작업이라면 따로 transaction을 붙일 필요가 없고, INSERT/UPDATE 작업을 하는 경우라면 원자성이 필요한 범위를 생각해서 붙이면 될 것 같습니다

f-lab-jd commented on src/main/java/dev/cass/knorda/api/product/controller/ProductController.java:
```diff
@@ -0,0 +1,81 @@
+
+	@DeleteMapping("/products/{productId}")
+	public ResponseEntity<Void> deleteProduct(@PathVariable int productId, HttpSession session) {
+		String loggedInUser = SessionManageUtils.getMemberName(session);
```
로그인유저를 불러오는 것이 중복이 되는데, 중복을 막기 위해서는 어떤 방법이 있을까요?
 (답변만 달아 주시고, 다음 PR에서 처리해주시면 좋을 것 같습니다)

Cass07 replied:
커스텀 어노테이션을 사용해서, 어노테이션 사용 시 로그인유저의 정보를 가져오는 `String loggedInUser = SessionManageUtils.getMemberName(session);` 의 값을 매개변수로 자동으로 넘겨주게 하는 방식이 있습니다.

f-lab-jd commented on src/main/java/dev/cass/knorda/api/product/facade/ProductFacade.java:
```diff
@@ -0,0 +1,63 @@
+	}
+
+	public ProductRegisterDto.RegisterResponse updateProduct(int productId,
+		ProductRegisterDto.RegisterRequest registerRequest,
```
coding convention 이  올바르게 적용되고 있나요~? 개행이 일관성있지 않은 부분이 있어서 여쭈어봅니다.

Cass07 replied:
save와 커밋 시 자동으로 code reformat이 진행되도록 플러그인을 세팅했습니다.

f-lab-jd commented on src/main/java/dev/cass/knorda/api/product/facade/ProductFacade.java:
```diff
@@ -0,0 +1,63 @@
+	public void deleteProduct(int productId, String loggedInUser) {
+		Product product = productService.findById(productId);
+		if (!product.getMember().getMemberName().equals(loggedInUser)) {
+			throw new NotYourProductException();
```
예외의 이름이 구체적이기는 하지만, 명료하지는 않습니다.
왜냐면 `You` 라는 표현이 비즈니스 로직에 쓰이기에는 모호한 부분이 있어서요 ㅎㅎ 
`loggedInUser 가 소유한 Product 가 아니다`  라는 표현이 드러났으면 좋겠습니다.

Cass07 replied:
명칭을 `ProductNotOwnedByLoggedInMemberException`로 변경했습니다.
또한 매개 변수 변수명에서 `User`, `Member` 단어가 혼용되지 않도록 `Member`로 변경하고, `SESSION_USER`도 `SESSION_MEMBER`로 변경하였습니다.

f-lab-jd commented on src/main/java/dev/cass/knorda/api/product/service/ProductService.java:
```diff
@@ -0,0 +1,80 @@
+		return productRepository.findFirstByProductId(productId).orElseThrow(ProductNotExistException::new);
+	}
+
+	public Product save(ProductRegisterDto.RegisterRequest registerRequest, Member member) {
```
`ProductFacade` 가 있는데, ProductService 도 DTO 에 대해서 알아야할까요~? 
나아가서 Member 도 알아야하나 라는 생각도 듭니다.

Cass07 replied:
Facade 레이어에서 DTO를 Entity로 변경하고, Service의 메소드의 매개변수로는 Entity를 넘겨주도록 변경하였습니다.

f-lab-jd commented on src/main/java/dev/cass/knorda/api/product/service/ProductService.java:
```diff
@@ -0,0 +1,80 @@
+	public Product save(ProductRegisterDto.RegisterRequest registerRequest, Member member) {
+		Product product = registerRequest.toEntity();
+		product.setMember(member);
+		return productRepository.saveAndFlush(product);
```
save 가 아니라 saveAndFlush 를 사용하신 이유는 무엇인가요?

Cass07 replied:
product의 name이 unique함을 보장하기 위해서, 중복 검증 로직을 사용하되 실제 Table에 락을 거는 것은 MySQL이 Unique Key를 INSERT/UPDATE할 때 수행하므로 Spring에서 락을 걸지 않고 MySQL에 위임하는 것으로 정하였습니다.
그렇다면, 동시성 문제로 인해서 검증 로직에 성공했으나 실제로 name이 unique하지 않았다면, INSERT/UPDATE가 실제로 쿼리되어야, `DataIntegrityViolationException`이 발생하면서 Spring에서 에러 여부를 알 수 있게 됩니다.
그러나 JPA는 영속성 컨텍스트의 존재로 인해서, JPA Repository save/update 메소드를 호출했을 때 영속성 컨텍스트에 객체의 변경 사항이 저장되고 바로 쿼리문이 나감을 보장하지 않습니다.
따라서 비즈니스 로직을 수행하면서 name이 unique하지 않음을 바로 알 수 있도록, 변경 사항에 대한 쿼리문을 바로 쿼리할 수 있도록 save 후에 flush를 바로 실행하는 saveAndFlush 메소드를 사용하였습니다.

f-lab-jd reviewed (commented):
고생하셨습니다! 리뷰드린 내용한번 확인 부탁드립니다. 
그리고 `Service` 와 구분되어서 `Facade` 의 역할과 책임에 대해서 한번 정의해보시면 좋을 것 같습니다. 그리고 `ProductFacade`라고 네이밍을 해주셨는데, 상황에따라(누가 짓느냐에 따라) `MemberFacade` 도 될 수 있는 것 아닌가요~?  

네이밍에 명확하게 일관성을 가져갈 수 있는 룰이 있으신거라면 말씀부탁드립니다!

Cass07 committed:
[Refactor] product specification의 분리

service 레이어에 있던 동적 쿼리를 작성하는 specification을 분리
- 기존에 Repository 클래스를 두 개 만들어서 하나는 구현체를 작성 후 상속시키는 방식으로 작성하려 했는데, 이 경우 JpaRepository의 모든 메소드를 구현해야 하는 것으로 보임
- Service 레이어가 Specification을 작성하는 책임을 지는 것은 맞지 않으므로, specification을 작성하는 메소드를 새로이 클래스를 작성하여 분리
- Criteria를 사용하여 동적 쿼리를 빌드할 때, Specification 클래스가 specification 생성을 담당하고, 이를 쿼리하는 것은 Repository 인터페이스(와 구현체)가 담당하도록 역할을 분리

Cass07 committed:
[Refactor] PR 리뷰 반영

PR 리뷰사항 수정
- INSERT/UPDATE service 메소드를 호출하는 Facade 메소드에 대해 Transactional 어노테이션 추가
- NotYourProductException 예외의 이름을 ProductNotOwnedByLoggedInMemberException 로 변경
- 매개 변수 및 일부 클래스 변수의 User 를 Member 로 변경
- DTO 의 핸들링은 가능한 한 Facade 레이어에서 하고, Service 에는 Entity 매개변수를 넘기도록 변경

f-lab-jd reviewed (approved):
👍

Cass07 committed:
[Refactor] PR 리뷰 반영

PR 리뷰사항 추가 수정
- JPA의 saveAndFlush를 save로 변경
- Facade에서는 비즈니스 로직을 묶는 것과 Entity와 DTO의 변환을 담당하도록 정의하고, Service에서는 Entity만 핸들링할 수 있도록 매개변수를 변경
- Update 로직을 JPA Save로 처리하도록 변경
- 모든 Facade 메소드에 대해 Transactinal 어노테이션 추가

Cass07 committed:
[Refactor] code smell 수정

Static method만을 포함하는 ProductSpecification 클래스에 private 생성자 추가
